# Parte 1

In [ ]:
# NOTA: ES NECESARIO RECORRER ESTA CELDA PARA CONTINUAR CON LAS DEMAS EN PARTE 1

# Suavizado Add-one (suavizado de Laplace): https://github.com/kapumota/Actividades-CC0C2/blob/main/Cuadernos-CC0C2/Clase2/Modelos-lenguaje1.ipynb
# Suavizado Add-k: https://github.com/kapumota/Actividades-CC0C2/blob/main/Cuadernos-CC0C2/Clase2/Modelos-lenguaje2.ipynb
# Suavizado Backoff:  https://github.com/kapumota/Actividades-CC0C2/blob/main/Cuadernos-CC0C2/Clase2/Counts-backoff-suavizado.ipynb
# Suavizado Stupid Backoff: https://github.com/kapumota/Actividades-CC0C2/blob/main/Cuadernos-CC0C2/Clase2/Topicos-avanzados.ipynb

corpus = "all models are wrong. a model is wrong. some models are useful"
vocab = {'<s>','</s>','a', 'all','are','model','models','some','useful','wrong'}

# Tokenizar
tokens = corpus.lower().replace('.', '').split()
tokens.insert(0, '<s>')
tokens.append('</s>')

# Frecuencias de bigramas y unigramas
bigram_counts = {}
unigram_counts = {}
for i in range(len(tokens) - 1):
  bigram = (tokens[i], tokens[i + 1])
  if bigram in bigram_counts:
    bigram_counts[bigram] += 1
  else:
    bigram_counts[bigram] = 1
  if tokens[i] in unigram_counts:
    unigram_counts[tokens[i]] += 1
  else:
    unigram_counts[tokens[i]] = 1

# Calcular el total de palabras (tokens)
total_words = sum(unigram_counts.values())



Este código tiene como objetivo procesar un corpus de texto para realizar un análisis de frecuencias de bigramas (pares de palabras consecutivas) y unigramas (palabras individuales).

***Definición del Corpus y Vocabulario:***

Se define un corpus de texto (corpus) y un conjunto de palabras para el vocabulario (vocab), lo cual es útil para limitar las palabras que se analizarán.

***Tokenización:***

Se convierte el corpus a minúsculas, se eliminan los puntos y se divide en palabras. Además, se insertan los tokens especiales `<s>` al inicio y `</s>` al final para marcar los límites de la frase.

***Cálculo de Frecuencias***

Se crean dos diccionarios: bigram_counts para almacenar la frecuencia de aparición de bigrams y unigram_counts para la frecuencia de unigrams.
A través de un bucle for, el código recorre la lista de tokens para calcular:

**Bigrams:** Pares consecutivos de palabras. Si un bigram ya existe en bigram_counts, su contador se incrementa en 1; si no, se crea una nueva entrada para ese bigram con un contador inicial de 1.

**Unigrams:** Cada palabra individual en la lista de tokens (excepto la última, que no se cuenta en este bucle). Se actualizan de manera similar a los bigrams.



## Calcule las probabilida de todos los bigramas sin suavizado

In [ ]:
# Probabilidades de bigramas
bigram_prob ={}
for bigram, contar in bigram_counts.items():
  # Encontrar la frecuencia de la primera palabra del bigrama
  prim_palabra= tokens.count(bigram[0])

  # Calcular la probabilidad del bigrama
  bigram_prob[bigram] = contar/prim_palabra

# Resultado
for bigram, prob in bigram_prob.items():
  print(f"P({bigram[1]}|{bigram[0]}) = {prob:.3f}")


P(all|<s>) = 1.000
P(models|all) = 1.000
P(are|models) = 1.000
P(wrong|are) = 0.500
P(a|wrong) = 0.500
P(model|a) = 1.000
P(is|model) = 1.000
P(wrong|is) = 1.000
P(some|wrong) = 0.500
P(models|some) = 1.000
P(useful|are) = 0.500
P(</s>|useful) = 1.000


Este código calcula las probabilidades de bigramas a partir de las frecuencias obtenidas previamente en la celda anterior.

Se crea un diccionario bigram_prob para almacenar las probabilidades condicionales de cada bigrama.

El código recorre cada bigrama en bigram_counts (donde se almacenaron las frecuencias de bigramas previamente).

Para cada bigrama `(palabra1, palabra2)`, se busca la frecuencia de la primera palabra palabra1 en la lista de tokens `(tokens.count(bigram[0]))`.

La salida representa las probabilidades condicionales de que una palabra aparezca después de otra en el corpus dado.

**OUTPUT:**

Las probabilidades de 1.000 indican que cada vez que se encuentra la primera palabra, siempre le sigue la segunda palabra.

Las probabilidades de 0.500 indican que la primera palabra puede ser seguida por dos opciones diferentes con igual probabilidad


## Calcule las probabilidades de todos los bigramas y el bigrama no visto "a models" con suavizado add-one

In [ ]:
# Probabilidades de bigramas con suavizado add-one

bigram_prob_addone = {}
for bigram in [(w1, w2) for w1 in vocab for w2 in vocab]:
  count_bigram = bigram_counts.get(bigram, 0)
  count_unigram = unigram_counts.get(bigram[0], 0)

  bigram_prob_addone[bigram] = (count_bigram + 1) / (count_unigram + len(vocab))

# Resultado
for bigram, prob in bigram_prob_addone.items():
  print(f"P({bigram[1]}|{bigram[0]}) = {prob:.3f}")


# Probabilidad del bigrama no visto "a models"

bigram_no_visto = ("a", "models")
prob_no_visto = bigram_prob_addone.get(bigram_no_visto, 0)
print("\nBigrama no visto (a models)")
print(f"\nP({bigram_no_visto[1]}|{bigram_no_visto[0]}) = {prob_no_visto:.3f}")

P(model|model) = 0.053
P(a|model) = 0.053
P(useful|model) = 0.053
P(all|model) = 0.053
P(are|model) = 0.053
P(wrong|model) = 0.053
P(some|model) = 0.053
P(models|model) = 0.053
P(<s>|model) = 0.053
P(</s>|model) = 0.053
P(model|a) = 0.105
P(a|a) = 0.053
P(useful|a) = 0.053
P(all|a) = 0.053
P(are|a) = 0.053
P(wrong|a) = 0.053
P(some|a) = 0.053
P(models|a) = 0.053
P(<s>|a) = 0.053
P(</s>|a) = 0.053
P(model|useful) = 0.053
P(a|useful) = 0.053
P(useful|useful) = 0.053
P(all|useful) = 0.053
P(are|useful) = 0.053
P(wrong|useful) = 0.053
P(some|useful) = 0.053
P(models|useful) = 0.053
P(<s>|useful) = 0.053
P(</s>|useful) = 0.105
P(model|all) = 0.053
P(a|all) = 0.053
P(useful|all) = 0.053
P(all|all) = 0.053
P(are|all) = 0.053
P(wrong|all) = 0.053
P(some|all) = 0.053
P(models|all) = 0.105
P(<s>|all) = 0.053
P(</s>|all) = 0.053
P(model|are) = 0.036
P(a|are) = 0.036
P(useful|are) = 0.071
P(all|are) = 0.036
P(are|are) = 0.036
P(wrong|are) = 0.071
P(some|are) = 0.036
P(models|are) = 0.036
P(<s>|are

El suavizado Add-One se utiliza para evitar probabilidades de cero en bigramas no vistos en el corpus. Entonces, si un bigrama no aparece en el conjunto de datos original, se le asigna una probabilidad pequeña en lugar de ser ignorado.

**OUTPUT**

Este resultado `P(models|a) = 0.053 ` indica que, según el modelo con suavizado, existe una probabilidad del 5.3% de que la palabra `models` siga a `a`, aunque este bigrama no haya sido visto en el corpus original.

Sin el suavizado, la probabilidad de `P(models|a) `sería 0 porque el bigrama `a models` no aparece en el texto. Esto puede ser un problema ya que resultaría en probabilidades nulas para palabras posibles. El suavizado evita este problema y asegura que todos los bigramas tengan al menos una probabilidad pequeña.

## Calcule las probabilidades de todos los bigramas y el bigrama no visto "a models" con suavizado add-k. Pruebe k=0.05 y k=0.15

In [ ]:
def prob_addk(tokens, vocab, k):

  # Probabilidades de bigramas con suavizado add-k
  bigram_prob_addk = {}
  for bigram in [(w1, w2) for w1 in vocab for w2 in vocab]:
    count_bigram = bigram_counts.get(bigram, 0)
    count_unigram = unigram_counts.get(bigram[0], 0)
    bigram_prob_addk[bigram] = (count_bigram + k) / (count_unigram + k * len(vocab))

  return bigram_prob_addk

# Calcular probabilidades con k=0.05
k005 = prob_addk(tokens, vocab, 0.05)

print("Probabilidades con suavizado add-k (k=0.05):\n")
for bigram, prob in k005.items():
  print(f"P({bigram[1]}|{bigram[0]}) = {prob:.3f}")


bigram_no_visto = ("a", "models")
prob005 = k005.get(bigram_no_visto, 0)
print("\nBigrama no visto (a models) con k=0.05")
print(f"\nP({bigram_no_visto[1]}|{bigram_no_visto[0]}) = {prob005:.3f}")

# Calcular probabilidades con k=0.15
k015 = prob_addk(tokens, vocab, 0.15)

print("\n\nProbabilidades con suavizado add-k (k=0.15):")
for bigram, prob in k015.items():
  print(f"P({bigram[1]}|{bigram[0]}) = {prob:.3f}")

bigram_no_visto = ("a", "models")
prob015 = k015.get(bigram_no_visto, 0)
print("\nBigrama no visto (a models) con k=0.15")
print(f"\nP({bigram_no_visto[1]}|{bigram_no_visto[0]}) = {prob015:.3f}")

Probabilidades con suavizado add-k (k=0.05):

P(model|model) = 0.005
P(a|model) = 0.005
P(useful|model) = 0.005
P(all|model) = 0.005
P(are|model) = 0.005
P(wrong|model) = 0.005
P(some|model) = 0.005
P(models|model) = 0.005
P(<s>|model) = 0.005
P(</s>|model) = 0.005
P(model|a) = 0.111
P(a|a) = 0.005
P(useful|a) = 0.005
P(all|a) = 0.005
P(are|a) = 0.005
P(wrong|a) = 0.005
P(some|a) = 0.005
P(models|a) = 0.005
P(<s>|a) = 0.005
P(</s>|a) = 0.005
P(model|useful) = 0.005
P(a|useful) = 0.005
P(useful|useful) = 0.005
P(all|useful) = 0.005
P(are|useful) = 0.005
P(wrong|useful) = 0.005
P(some|useful) = 0.005
P(models|useful) = 0.005
P(<s>|useful) = 0.005
P(</s>|useful) = 0.111
P(model|all) = 0.005
P(a|all) = 0.005
P(useful|all) = 0.005
P(all|all) = 0.005
P(are|all) = 0.005
P(wrong|all) = 0.005
P(some|all) = 0.005
P(models|all) = 0.111
P(<s>|all) = 0.005
P(</s>|all) = 0.005
P(model|are) = 0.003
P(a|are) = 0.003
P(useful|are) = 0.057
P(all|are) = 0.003
P(are|are) = 0.003
P(wrong|are) = 0.057
P(som

El suavizado Add-k modifica el cálculo de probabilidades al sumar un valor `k` a los conteos de cada bigrama. Esto asegura que incluso los bigramas que no se han observado en el corpus reciban una pequeña probabilidad en lugar de cero.

**OUTPUT**

`P(models|a) = 0.005, K=0.05`

Esta probabilidad indica que, cuando se utiliza un valor pequeño de k=0.05, el modelo asigna una probabilidad muy baja a que la palabra `models` siga a `a`.
Dado que el bigrama `a models` no aparece en el corpus original, el suavizado add-k evita que la probabilidad sea cero y asigna un valor positivo pequeño.

`P(models|a) = 0.014, K=0.15`

Al usar un valor más grande de k=0.15, la probabilidad asignada a `P(models|a)` aumenta a `1.4%`. El incremento en k permite que el modelo dé más peso a los bigramas no observados, elevando sus probabilidades.

Podemos concluir que incrementar `k` asigna mayores probabilidades a bigramas no vistos siendo más flexible y menos restringido a los datos originales del corpus.


## Calcule las probabilidades de todos los bigramas y el bigrama no visto "a models" con back-off y stupid-backoff

In [ ]:
def backoff_bigram_probability(w1, w2, unigram_counts, bigram_counts, total_words):
    bigram_count = bigram_counts.get((w1, w2), 0)
    if bigram_count > 0:
        return bigram_count / unigram_counts.get(w1, 0)
    else:
        # Back-off to unigram probability
        return unigram_counts.get(w2, 0) / total_words

# Prob para bigramas
for w1 in vocab:
    for w2 in vocab:
        prob = backoff_bigram_probability(w1, w2, unigram_counts, bigram_counts, total_words)
        if prob > 0:
            print(f"P({w2}|{w1}) = {prob:.3f}")

# Prob para "a models"
probrar_backoff = backoff_bigram_probability('a', 'models', unigram_counts, bigram_counts, total_words)
print(f"P(models|a) con back-off = {probrar_backoff:.3f}")


P(model|model) = 0.077
P(useful|model) = 0.077
P(<s>|model) = 0.077
P(some|model) = 0.077
P(a|model) = 0.077
P(wrong|model) = 0.154
P(models|model) = 0.154
P(are|model) = 0.154
P(all|model) = 0.077
P(model|useful) = 0.077
P(useful|useful) = 0.077
P(<s>|useful) = 0.077
P(some|useful) = 0.077
P(a|useful) = 0.077
P(wrong|useful) = 0.154
P(models|useful) = 0.154
P(are|useful) = 0.154
P(all|useful) = 0.077
P(</s>|useful) = 1.000
P(model|<s>) = 0.077
P(useful|<s>) = 0.077
P(<s>|<s>) = 0.077
P(some|<s>) = 0.077
P(a|<s>) = 0.077
P(wrong|<s>) = 0.154
P(models|<s>) = 0.154
P(are|<s>) = 0.154
P(all|<s>) = 1.000
P(model|some) = 0.077
P(useful|some) = 0.077
P(<s>|some) = 0.077
P(some|some) = 0.077
P(a|some) = 0.077
P(wrong|some) = 0.154
P(models|some) = 1.000
P(are|some) = 0.154
P(all|some) = 0.077
P(model|a) = 1.000
P(useful|a) = 0.077
P(<s>|a) = 0.077
P(some|a) = 0.077
P(a|a) = 0.077
P(wrong|a) = 0.154
P(models|a) = 0.154
P(are|a) = 0.154
P(all|a) = 0.077
P(model|wrong) = 0.077
P(useful|wrong) = 

El código implementa un enfoque de back-off para calcular las probabilidades de bigramas, utilizando una estrategia que recurre a las probabilidades de unigramas cuando el bigrama no se ha observado en el corpus.

**OUTPUT**

El código genera las probabilidades para todos los posibles bigramas en el vocabulario con el método back-off

`P(models|a) con back-off = 0.154`

Dado que el bigrama `a models` no se ha observado en el corpus, el cálculo se respalda en la probabilidad del unigrama P(models). La probabilidad resultante de 0.154 muestra que, aunque `models `no aparece directamente después de `a` en el corpus original, la frecuencia de models es suficiente para que el modelo le asigne una probabilidad significativa en combinación con a.


In [ ]:
# Stupid backoff
def stupid_backoff_bigram_probability(w1, w2, unigram_counts, bigram_counts, total_words, factor_lamda=0.4):
    bigram_count = bigram_counts.get((w1, w2), 0)
    if bigram_count > 0:
        return bigram_count / unigram_counts.get(w1, 0)
    else:
        return factor_lamda * (unigram_counts.get(w2, 0) / total_words)

# Calculate probabilidad for all bigrams
for w1 in vocab:
    for w2 in vocab:
        prob = stupid_backoff_bigram_probability(w1, w2, unigram_counts, bigram_counts, total_words)
        if prob > 0:
            print(f"P({w2}|{w1}) = {prob:.3f}")

# Calculate probability for unseen bigram "a models"
probrar_backoff_tonto = stupid_backoff_bigram_probability('a', 'models', unigram_counts, bigram_counts, total_words)
print(f"P(models|a) con stupid back-off = {probrar_backoff_tonto:.3f}")

P(model|model) = 0.031
P(useful|model) = 0.031
P(<s>|model) = 0.031
P(some|model) = 0.031
P(a|model) = 0.031
P(wrong|model) = 0.062
P(models|model) = 0.062
P(are|model) = 0.062
P(all|model) = 0.031
P(model|useful) = 0.031
P(useful|useful) = 0.031
P(<s>|useful) = 0.031
P(some|useful) = 0.031
P(a|useful) = 0.031
P(wrong|useful) = 0.062
P(models|useful) = 0.062
P(are|useful) = 0.062
P(all|useful) = 0.031
P(</s>|useful) = 1.000
P(model|<s>) = 0.031
P(useful|<s>) = 0.031
P(<s>|<s>) = 0.031
P(some|<s>) = 0.031
P(a|<s>) = 0.031
P(wrong|<s>) = 0.062
P(models|<s>) = 0.062
P(are|<s>) = 0.062
P(all|<s>) = 1.000
P(model|some) = 0.031
P(useful|some) = 0.031
P(<s>|some) = 0.031
P(some|some) = 0.031
P(a|some) = 0.031
P(wrong|some) = 0.062
P(models|some) = 1.000
P(are|some) = 0.062
P(all|some) = 0.031
P(model|a) = 1.000
P(useful|a) = 0.031
P(<s>|a) = 0.031
P(some|a) = 0.031
P(a|a) = 0.031
P(wrong|a) = 0.062
P(models|a) = 0.062
P(are|a) = 0.062
P(all|a) = 0.031
P(model|wrong) = 0.031
P(useful|wrong) = 

El código utiliza stupid backoff, que es una variante simplificada del backoff tradicional en modelos de lenguaje.

**OTUPUT:**

`P(models|a) con stupid back-off = 0.062`

El resultado 0.062 indica que la probabilidad del bigrama no visto `a models `es del 6.2%. Esta probabilidad es calculada multiplicando la probabilidad del unigrama `P(models)` por el factor` λ (0.4)`. Esto refleja la confianza reducida que tiene el modelo en la combinación no vista en comparación con un bigrama observado.



## Conclusiones parte 1

La variación en las probabilidades `P(models|a)` se debe a cómo cada método maneja la parte de bigramas no vistos. **Add-one y add-k** suavizan asignando pequeñas probabilidades uniformementen, mientras que, **Back-off y stupid backoff** usan la frecuencia general de models, dando más flexibilidad y ajustando las probabilidades según la proporcion de unigramas. Esto causa que métodos como back-off asignen probabilidades más altas comparado con los suavizados, que distribuyen el peso de forma más pareja.

# Parte 2

In [ ]:
# NOTA: ES NECESARIO RECORRER ESTA CELDA PARA CONTINUAR CON LAS DEMAS EN LA PARTE 2

# Suavizado Good-Turing: https://github.com/kapumota/Actividades-CC0C2/blob/main/Cuadernos-CC0C2/Clase2/Modelos-lenguaje2.ipynb
# Maxima verosimilitud https://github.com/kapumota/Actividades-CC0C2/blob/main/Cuadernos-CC0C2/Clase2/Modelos-lenguaje2.ipynb

corpus = "all models are wrong. a model is wrong. some models are useful"
vocab = {'<s>','</s>','<UNK>','a', 'all','are','model','models','some','useful','wrong'}

# Tokenizar
tokens = corpus.lower().replace('.', '').split()
tokens.insert(0, '<s>')
tokens.append('</s>')

# Reemplazamos palabras desconocidas por <UNK>
tokens = [token if token in vocab else '<UNK>' for token in tokens]



**Reemplazo de Palabras Desconocidas (`<UNK>`):**

Cada token se verifica en el vocabulario, si una palabra no está en el conjunto vocab, se reemplaza por `<UNK>`.

## Good Turing

In [ ]:
from collections import Counter

# Cálculo de unigramas
unigram_counts = Counter(tokens)

# Calculamos r (frecuencia de unigramas) y N_r (número de unigramas con frecuencia r)
count_of_counts = Counter(unigram_counts.values())

# Mostramos los resultados
print("Conteos de unigramas (r):", unigram_counts)
print("Conteo de frecuencias N_r:", count_of_counts)

# Good-Turing
def good_turing_counts(unigram_counts):
    count_of_counts = Counter(unigram_counts.values())
    adjusted_counts = {}

    for unigram, count in unigram_counts.items():
        if count + 1 in count_of_counts:
            adjusted_counts[unigram] = (count + 1) * (count_of_counts[count + 1] / count_of_counts[count])
        else:
            adjusted_counts[unigram] = count

    return adjusted_counts

# Aplicamos Good-Turing a los unigramas
adjusted_unigram_counts = good_turing_counts(unigram_counts)

# Mostramos los resultados ajustados
print("Conteos ajustados con Good-Turing:", adjusted_unigram_counts)


Conteos de unigramas (r): Counter({'models': 2, 'are': 2, 'wrong': 2, '<s>': 1, 'all': 1, 'a': 1, 'model': 1, '<UNK>': 1, 'some': 1, 'useful': 1, '</s>': 1})
Conteo de frecuencias N_r: Counter({1: 8, 2: 3})
Conteos ajustados con Good-Turing: {'<s>': 0.75, 'all': 0.75, 'models': 2, 'are': 2, 'wrong': 2, 'a': 0.75, 'model': 0.75, '<UNK>': 0.75, 'some': 0.75, 'useful': 0.75, '</s>': 0.75}


El método Good Turing primero, calcula cuántas veces aparece cada palabra (unigrama) en el corpus. Luego, determina cuántos unigramas tienen exactamente la misma frecuencia, agrupando los conteos para ver cuántas palabras aparecen solo una vez o dos veces.

Para ajustar los conteos, el método analiza las palabras que ocurren solo una vez y reduce su conteo original, asegurando que las probabilidades reflejen la posibilidad de que haya palabras que no se hayan visto en el corpus. Si una palabra tiene una frecuencia baja (por ejemplo, aparece solo una vez), el método ajusta ese valor hacia abajo, mientras que mantiene los conteos originales para palabras más comunes. De esta forma, Good-Turing reasigna parte del conteo de las palabras observadas a posibles palabras que podrían no haber aparecido, mejorando la precisión del modelo en casos de datos escasos.

**OUTPUT**

`Counter({'models': 2, 'are': 2, 'wrong': 2, '<s>': 1, 'all': 1, 'a': 1, 'model': 1, '<UNK>': 1, 'some': 1, 'useful': 1, '</s>': 1})`

Esto muestra que models, are, y wrong aparecen 2 veces cada uno, mientras que todos los demás unigramas, incluidos los tokens <s> y </s>, aparecen solo una vez.

`Counter({1: 8, 2: 3})`

Hay 8 unigramas que aparecen solo una vez y 3 unigramas que aparecen dos veces en el corpus. Esto indica que la mayoría de los unigramas son raros.

`{'<s>': 0.75, 'all': 0.75, 'models': 2, 'are': 2, 'wrong': 2, 'a': 0.75, 'model': 0.75, '<UNK>': 0.75, 'some': 0.75, 'useful': 0.75, '</s>': 0.75}`

Los unigramas quetenían un conteo de 1 se ajustan a 0.75. Esto se debe a que el método Good-Turing redistribuye el conteo para reflejar la posibilidad de eventos no observados reduciendo ligeramente los conteos de palabras raras.

Los unigramas que ya tenían un conteo de 2 se mantienen sin cambios porque el ajuste no se aplica en casos donde hay suficiente frecuencia.




## (f) **Para** r < 3 calcular las probabilidades de todos los unigramas



In [ ]:
from collections import Counter

# Good-Turing
def good_turing_counts(ngram_counts):
    count_of_counts = Counter(ngram_counts.values())
    adjusted_counts = {}

    for ngram, count in ngram_counts.items():
        if count + 1 in count_of_counts:
            adjusted_counts[ngram] = (count + 1) * (count_of_counts[count + 1] / count_of_counts[count])
        else:
            adjusted_counts[ngram] = count

    return adjusted_counts

# Cálculo de unigramas
unigram_counts = Counter(tokens)

# Aplicar suavizado Good-Turing
adjusted_unigram_counts = good_turing_counts(unigram_counts)

# Calcular probabilidades para r < 3
total_tokens = sum(unigram_counts.values())
probabilidad = {}

for unigram, adjusted_count in adjusted_unigram_counts.items():
    if unigram_counts[unigram] < 3:  # r < 3
        probabilidad[unigram] = adjusted_count / total_tokens

# Mostrar resultados
print("Probabilidades para unigramas con r < 3:")
for unigram, prob in probabilidad.items():
    print(f"Unigrama: {unigram}, Probabilidad: {prob:.4f}")


Probabilidades para unigramas con r < 3:
Unigrama: <s>, Probabilidad: 0.0536
Unigrama: all, Probabilidad: 0.0536
Unigrama: models, Probabilidad: 0.1429
Unigrama: are, Probabilidad: 0.1429
Unigrama: wrong, Probabilidad: 0.1429
Unigrama: a, Probabilidad: 0.0536
Unigrama: model, Probabilidad: 0.0536
Unigrama: <UNK>, Probabilidad: 0.0536
Unigrama: some, Probabilidad: 0.0536
Unigrama: useful, Probabilidad: 0.0536
Unigrama: </s>, Probabilidad: 0.0536


El código filtra los unigramas que tienen una frecuencia (r) menor a 3. Solo se calculan las probabilidades para estos unigramas.
La probabilidad de cada unigrama se obtiene dividiendo su conteo ajustado por el total de tokens en el corpus. Esto asegura que las probabilidades estén normalizadas y sumen 1.

**OUTPUT**

La mayoría de los unigramas aparecen solo una vez (r = 1) y sus probabilidades ajustadas son alrededor de 0.0536. Esto muestra cómo el método Good-Turing reduce  sus conteos reales para tener algo de probabilidad para posibles eventos no observados.

Las palabras que aparecen más de una vez, como `models`, `are` y `wrong`, tienen probabilidades más altas (0.1429). Esto indica que Good-Turing mantiene sus conteos ajustados cercanos a los valores originales, ya que hay suficiente información para estimar su frecuencia con más precisión.

##  (g) Calcular probabilidades de unigramas de la parte 1 que aparecen con mayor frecuencia $ r= 3$

In [ ]:
from collections import Counter

# Cálculo de unigramas
unigram_counts = Counter(tokens)

# Total de tokens en el corpus
total_tokens = sum(unigram_counts.values())

# Determinar el valor máximo de r (frecuencia)
max_r = max(unigram_counts.values())

# Probabilidad con MLE para unigramas con r = max_r (r = 2) ya que no hay palabras que se repitan 3 veces en este corpus
probabilidad_mle = {}

for unigram, count in unigram_counts.items():
    if count == max_r:  # Solo para los unigramas con la frecuencia máxima
        probabilidad_mle[unigram] = count / total_tokens

# Mostrar resultados
print(f"Probabilidad estimada por MLE para unigramas con r = {max_r}:")
for unigram, prob in probabilidad_mle.items():
    print(f"Unigrama: {unigram}, Probabilidad: {prob:.4f}")


Probabilidad estimada por MLE para unigramas con r = 2:
Unigrama: models, Probabilidad: 0.1429
Unigrama: are, Probabilidad: 0.1429
Unigrama: wrong, Probabilidad: 0.1429


Se identifica la frecuencia máxima `r` en el corpus. En este caso, el valor
máximo de `r` es 2, ya que no hay palabras que se repitan 3 veces o más. Para los unigramas que tienen la frecuencia máxima (r = 2), la probabilidad se calcula dividiendo el conteo de cada unigrama entre el total de tokens en el corpus.

**OUTPUT**

Las palabras `models`, `are` y `wrong` tienen la frecuencia máxima `r=2` en el corpus. Cada una aparece 2 veces. Usando MLE, la probabilidad de cada una de estas palabras es `0.1429`, lo que se calcula dividiendo 2 (frecuencia de la palabra) entre 14 (total de tokens). Estas probabilidades indican que hay un `14.29%` de posibilidad de que cualquiera de estas palabras aparezca en una posición dada.

## Demostrar que la suma de prob en (f) y (g) no es 1. Normaliza las probabilidades

In [ ]:
from collections import Counter

# Good-Turing
def good_turing_counts(ngram_counts):
    count_of_counts = Counter(ngram_counts.values())
    adjusted_counts = {}

    for ngram, count in ngram_counts.items():
        if count + 1 in count_of_counts:
            adjusted_counts[ngram] = (count + 1) * (count_of_counts[count + 1] / count_of_counts[count])
        else:
            adjusted_counts[ngram] = count

    return adjusted_counts

# Cálculo de unigramas
unigram_counts = Counter(tokens)

# Total de tokens en el corpus
total_tokens = sum(unigram_counts.values())

# Aplicar suavizado Good-Turing para r < 3
adjusted_unigram_counts = good_turing_counts(unigram_counts)

# Calcular las probabilidades con Good-Turing para r < 3
probabilidad_f = {}
for unigram, adjusted_count in adjusted_unigram_counts.items():
    if unigram_counts[unigram] < 3:  # r < 3
        probabilidad_f[unigram] = adjusted_count / total_tokens

# Probabilidades con MLE para unigramas con r = 2
max_r = 2
probabilidad_g = {}
for unigram, count in unigram_counts.items():
    if count == max_r:  # Solo para los unigramas con la frecuencia máxima
        probabilidad_g[unigram] = count / total_tokens

# Sumar las probabilidades de (f) y (g)
prob_total = sum(probabilidad_f.values()) + sum(probabilidad_g.values())

# Mostrar resultados
print("Probabilidades en (f):", probabilidad_f)
print("Probabilidades en (g):", probabilidad_g)
print(f"Suma total de las probabilidades en (f) y (g): {prob_total:.4f}")

# Normalizar las probabilidades si la suma no es 1
if prob_total != 1.0:
    normalizacion = 1 / prob_total
    probabilidad_f = {unigram: prob * normalizacion for unigram, prob in probabilidad_f.items()}
    probabilidad_g = {unigram: prob * normalizacion for unigram, prob in probabilidad_g.items()}
    prob_normalizada = sum(probabilidad_f.values()) + sum(probabilidad_g.values())
    print(f"Suma total normalizada de las probabilidades: {prob_normalizada:.4f}")


Probabilidades en (f): {'<s>': 0.05357142857142857, 'all': 0.05357142857142857, 'models': 0.14285714285714285, 'are': 0.14285714285714285, 'wrong': 0.14285714285714285, 'a': 0.05357142857142857, 'model': 0.05357142857142857, '<UNK>': 0.05357142857142857, 'some': 0.05357142857142857, 'useful': 0.05357142857142857, '</s>': 0.05357142857142857}
Probabilidades en (g): {'models': 0.14285714285714285, 'are': 0.14285714285714285, 'wrong': 0.14285714285714285}
Suma total de las probabilidades en (f) y (g): 1.2857
Suma total normalizada de las probabilidades: 1.0000


El código suma las probabilidades calculadas en (f) y (g). En este caso, la suma total es 1.2857, lo que indica que las probabilidades no están normalizadas (la suma debe ser exactamente 1 para ser válidas en un modelo de probabilidad).

Para corregir esto, se normalizan las probabilidades dividiendo cada una por el total (1.2857) para que la suma sea 1. Este proceso asegura que el modelo sea consistente y cumpla con las propiedades de las distribuciones de probabilidad.

**OUTPUT**

La variación se da porque Good-Turing redistribuye probabilidades para eventos raros, mientras que MLE asigna directamente a eventos frecuentes sin ajustes. Cuando se suman, las probabilidades exceden en 1, lo que requiere normalización para asegurar que el total unidad.

## USAR N5 y el valor original de los recuentos de frecuencia para calcular las probabilidades de todas las unigramas

In [ ]:
from collections import Counter
import math

# Good-Turing con N5
def good_turing_counts(ngram_counts, adjusted_N5):
    count_of_counts = Counter(ngram_counts.values())

    count_of_counts[5] = adjusted_N5

    adjusted_counts = {}

    for ngram, count in ngram_counts.items():
        if count + 1 in count_of_counts:
            adjusted_counts[ngram] = (count + 1) * (count_of_counts[count + 1] / count_of_counts[count])
        else:
            adjusted_counts[ngram] = count

    return adjusted_counts

# Corpus actualizado con la nueva oración
corpus = "all models are wrong. a model is wrong wrong wrong wrong. some models are useful"
vocab = {'<s>', '</s>', '<UNK>', 'a', 'all', 'are', 'model', 'models', 'some', 'useful', 'wrong'}

# Tokenización
tokens = corpus.lower().replace('.', '').split()
tokens.insert(0, '<s>')
tokens.append('</s>')

# Reemplazamos las palabras fuera del vocabulario por <UNK>
tokens = [token if token in vocab else '<UNK>' for token in tokens]

# Cálculo de unigramas
unigram_counts = Counter(tokens)

# Valor ajustado de N_5
# Utilizamos una aproximación lineal sencilla, un valor razonable podría ser cercano a 0.5
adjusted_N5 = 0.5

# Aplicamos Good-Turing con el valor ajustado de N_5
adjusted_unigram_counts = good_turing_counts(unigram_counts, adjusted_N5)

# Total de tokens en el corpus
total_tokens = sum(unigram_counts.values())

# Calcular las probabilidades usando los conteos ajustados
probabilidades= {unigram: adjusted_count / total_tokens for unigram, adjusted_count in adjusted_unigram_counts.items()}

# Mostrar resultados
print("Probabilidades con N_5 suavizado:")
for unigram, prob in probabilidades.items():
    print(f"Unigrama: {unigram}, Probabilidad: {prob:.4f}")


Probabilidades con N_5 suavizado:
Unigrama: <s>, Probabilidad: 0.0294
Unigrama: all, Probabilidad: 0.0294
Unigrama: models, Probabilidad: 0.1176
Unigrama: are, Probabilidad: 0.1176
Unigrama: wrong, Probabilidad: 0.2941
Unigrama: a, Probabilidad: 0.0294
Unigrama: model, Probabilidad: 0.0294
Unigrama: <UNK>, Probabilidad: 0.0294
Unigrama: some, Probabilidad: 0.0294
Unigrama: useful, Probabilidad: 0.0294
Unigrama: </s>, Probabilidad: 0.0294


El código usa una aproximación lineal para suavizar N5, Aquí se utiliza un valor de 0.5, evitando que N5 sea cero y se actualizó el corpus para que la palabra wrong aparezca 4 veces en una frase, asegurando que `N4 = 1`

**OUTPUT**

La palabra `wrong` tiene una probabilidad de `0.2941` porque aparece 4 veces en el corpus, lo que indica que es la palabra más frecuente.
Este ajuste se ve influenciado por el uso de Good-Turing y la consideración del valor suavizado de N5. Al asignar un valor aproximado de 0.5 a N5 , el modelo evita dar probabilidades de cero a eventos casi nulos, como eventos que podrían aparecer 5 veces. Esto permite al modelo manejar la falta de datos

# Parte 3

**Corpus**

Para este ejercicio, hemos utilizado un corpus de texto en español basado en un archivo llamado infopankki.en-es.es. Este archivo tiene como contenido la pagina de tour de Finlandia

**Descarga y Extracción**

Para comenzar, se carga el archivo infopankki.en-es.es. Este archivo es leído línea por línea utilizando la función leer_archivo. Posteriormente, se toma una muestra aleatoria de 1000 líneas para crear un conjunto de datos manejable.

**Tokenización**

El código convierte cada línea de texto a minúsculas y reemplaza cualquier puntuación con espacios. Esto se realiza mediante la función tokenizar, que también divide el texto en palabras individuales (tokens). Se maneja la puntuación y la conversión entre mayúsculas y minúsculas para estandarizar el texto.

**Lematización/Stemming**

La función lematizar se encarga de simplificar las palabras a su forma base. En este código, se eliminan los sufijos  como "s" para derivar una forma raíz. Ayuda a reducir la variabilidad en el texto al consolidar diferentes formas de una misma palabra.

**Filtrado de Palabras Raras**

Para reducir el tamaño del vocabulario y eliminar el ruido, el código filtra las palabras que aparecen menos de 5 veces en el corpus. Esto asegura que solo se mantengan las palabras más relevantes y frecuentes.

Construcción del Vocabulario

Después del preprocesamiento, se utiliza la función `construir_vocabulario` para generar un diccionario que asigna un identificador único a cada palabra en el corpus. Esto crea un vocabulario que se utilizará para entrenar modelos posteriores.

In [2]:
# Obtenido de https://github.com/Kinartb/CC0C2/blob/main/PC1/pc1_nlp.py
# Word2vec y GloVe:  https://github.com/kapumota/Actividades-CC0C2/blob/main/Cuadernos-CC0C2/Clase4/Embeddings.ipynb
import math
import collections
from typing import List, Tuple, Dict
import random
import numpy as np
import nltk
nltk.download('punkt')
from nltk.tokenize import word_tokenize
import re

#########
# Stopwords
from nltk.corpus import stopwords
nltk.download('stopwords')
stop_words = set(stopwords.words('spanish'))
#########
# Seleccion y prepacion del corpus

def leer_archivo(ruta):
    with open(ruta, 'r', encoding='utf-8') as file:
        return file.readlines()

# Usamos el archivo infopankki.en-es.es

ruta = 'infopankki.en-es.es'
archivo = leer_archivo(ruta)

# Se toma una mustra aleatorio de 1000 lineas del corpus original


def preprocesar_archivo(archivo, num_muestras=1000, umbral_frecuencia=5):
    documentos = []
    # Distribucion de frecuencia del corpus
    todas_palabras = []
    for muestra in random.sample(archivo, num_muestras):
        tokens = tokenizar(muestra)
        todas_palabras.extend(tokens)

    distribucion_frecuencia = nltk.FreqDist(todas_palabras)

    for muestra in random.sample(archivo, num_muestras):
        # Tokenizacion
        tokens = tokenizar(muestra)
        # Lemantizacion/Stemming
        tokens = [Lematizar(token) for token in tokens]
        # Remocion stopwords
        tokens = [token for token in tokens if token not in stop_words]
        # Filtrado de palabras raras
        tokens = [token for token in tokens if distribucion_frecuencia[token] >= umbral_frecuencia]
        documentos.append(tokens)
    return documentos

############
# Tokenizacion

def tokenizar(texto):
  # Convertir a minúsculas.
  texto = texto.lower()
  # Reemplazar la puntuación con espacios
  texto = re.sub(r'[^\w\s]', ' ', texto)
  # Dividir el texto en palabras
  tokens = texto.split()
  return tokens
#############
# Lemantizacion/Stemming
def Lematizar(token):
    if token.endswith('s'):
        return token[:-1]
    return token
##############
# Fin de preprocesamiento

#############

##############
corpus_entrenamiento = preprocesar_archivo(archivo)
print(f"Tamaño del corpus de entrenamiento: {len(corpus_entrenamiento)}")

def construir_vocabulario(corpus: List[List[str]]) -> Dict[str, int]:
    vocabulario = {}
    for documento in corpus:
        for token in documento:
            if token not in vocabulario:
                vocabulario[token] = len(vocabulario)
    return vocabulario

vocabulario = construir_vocabulario(corpus_entrenamiento)
tam_vocabulario = len(vocabulario)
print(f"Tamaño del vocabulario: {tam_vocabulario}")


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


Tamaño del corpus de entrenamiento: 1000
Tamaño del vocabulario: 220


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


Al finalizar el proceso, se obtiene el tamaño del corpus y el vocabulario generado donde:

Tamaño del corpus de entrenamiento: Muestra la cantidad de documentos (líneas de texto tokenizadas) procesadas.

Tamaño del vocabulario: Cuántas palabras únicas quedaron después del filtrado.

## Brown Clustering

### Brown Clustering simple

**Inicialización**

Se asigna cada palabra a su propio cluster inicial utilizando la función `inicializar_clusters`. Esto crea un diccionario donde cada palabra se asigna a un ID único de cluster, permitiendo que cada palabra comience como su propio cluster independiente.

**Cálculo de Probabilidades**

Para estimar las probabilidades de los clusters se calcula` P(c)` y `P(c_i,c_j)` que representa la probabilidad de cada cluster basado en la frecuencia de aparición de sus palabras en el corpus procesado. Se realiza en la función `calcular_probabilidades_Pc`, que determina qué tan común es cada cluster.

**Búsqueda de Fusión Óptima**

Se selecciona el par de clusters con el mínimo `ΔI` calculado y se realiza su fusión. Después de cada fusión, se actualizan las probabilidades `P(c)` para reflejar los cambios en los clusters. Este proceso se repite hasta alcanzar el número de cluester.


**Fusión de Clusters**

Se selecciona el par de clusters con el mínimo ΔI calculado y se realiza su fusión. Después de cada fusión, se actualizan las probabilidades `P(c)` para reflejar los cambios en los clusters. Este proceso se repite hasta alcanzar el número deseado de clusters, en este caso 100.

**Repetición**

El proceso de fusión continúa hasta que se alcanzan los `100` clusters deseados. Cada vez que se realiza una fusión, el código imprime el número de clusters restantes.

In [ ]:
import collections
import math
from typing import Dict, List, Tuple

##########
# Inicialización (Asignar cada palabra a su propio cluster)

def inicializar_clusters(vocabulario: Dict[str, int]) -> Dict[int, List[str]]:
    clusters = {i: [palabra] for palabra, i in vocabulario.items()}
    return clusters

clusters = inicializar_clusters(vocabulario)
print(f"Número inicial de clusters: {len(clusters)}")

##########
# Cálculo de probabilidades P(c)

def calcular_probabilidades_Pc(clusters: Dict[int, List[str]], corpus: List[List[str]]) -> Dict[int, float]:
    total_palabras = sum(len(doc) for doc in corpus)
    P_c = {}

    # Contar frecuencia de clases
    for c_id, palabras in clusters.items():
        frecuencia_cluster = sum(1 for doc in corpus for token in doc if token in palabras)
        P_c[c_id] = frecuencia_cluster / total_palabras

    return P_c

P_c = calcular_probabilidades_Pc(clusters, corpus_entrenamiento)
print("Probabilidades iniciales calculadas.")

##########
# Fusión de clusters

def fusionar_clusters_simple(clusters: Dict[int, List[str]], P_c: Dict[int, float], corpus: List[List[str]], num_clusters_deseado: int, epsilon: float = 1e-10):
    total_palabras = sum(len(doc) for doc in corpus)

    while len(clusters) > num_clusters_deseado:
        min_delta_I = None
        pair_to_merge = None

        cluster_ids = list(clusters.keys())
        for idx_i in range(len(cluster_ids)):
            c_i = cluster_ids[idx_i]
            palabras_i = set(clusters[c_i])
            for idx_j in range(idx_i + 1, len(cluster_ids)):
                c_j = cluster_ids[idx_j]
                palabras_j = set(clusters[c_j])

                # Calculamos P(c_i, c_j)
                frecuencia_conjunta = sum(
                    1 for doc in corpus if any(token in palabras_i for token in doc) and any(token in palabras_j for token in doc)
                )
                if frecuencia_conjunta == 0:
                    continue

                prob_conjunta = frecuencia_conjunta / total_palabras
                prob_conjunta = max(prob_conjunta, epsilon)
                P_ci = max(P_c[c_i], epsilon)
                P_cj = max(P_c[c_j], epsilon)

                delta_I = prob_conjunta * math.log(prob_conjunta / (P_ci * P_cj))

                if min_delta_I is None or delta_I < min_delta_I:
                    min_delta_I = delta_I
                    pair_to_merge = (c_i, c_j)

        if pair_to_merge is None:
            break  # No hay más fusiones posibles

        c_i, c_j = pair_to_merge

        # Fusionamos los clusters
        clusters[c_i].extend(clusters.pop(c_j))

        # Actualizamos P(c)
        P_c[c_i] += P_c.pop(c_j)

        print(f"Fusión realizada. Clusters restantes: {len(clusters)}")

##########
# Número de clusters deseado (100 clusters)
num_clusters_deseado = 100

fusionar_clusters_simple(clusters, P_c, corpus_entrenamiento, num_clusters_deseado)

print(f"Número final de clusters: {len(clusters)}")


Número inicial de clusters: 229
Probabilidades iniciales calculadas.
Fusión realizada. Clusters restantes: 228
Fusión realizada. Clusters restantes: 227
Fusión realizada. Clusters restantes: 226
Fusión realizada. Clusters restantes: 225
Fusión realizada. Clusters restantes: 224
Fusión realizada. Clusters restantes: 223
Fusión realizada. Clusters restantes: 222
Fusión realizada. Clusters restantes: 221
Fusión realizada. Clusters restantes: 220
Fusión realizada. Clusters restantes: 219
Fusión realizada. Clusters restantes: 218
Fusión realizada. Clusters restantes: 217
Fusión realizada. Clusters restantes: 216
Fusión realizada. Clusters restantes: 215
Fusión realizada. Clusters restantes: 214
Fusión realizada. Clusters restantes: 213
Fusión realizada. Clusters restantes: 212
Fusión realizada. Clusters restantes: 211
Fusión realizada. Clusters restantes: 210
Fusión realizada. Clusters restantes: 209
Fusión realizada. Clusters restantes: 208
Fusión realizada. Clusters restantes: 207
Fusión 

### Brown Clustering con mejoras en complejidad



**Búsqueda de Fusión Óptima**

Para determinar qué clusters deben fusionarse, se utiliza una cola de prioridad (`heap`) para almacenar los valores de `ΔI` para cada par de clusters. Esta métrica calcula la disminución en la información mutua cuando se consideran las probabilidades `P(c_i,c_j)` conjuntas realizando así más fácil la selección del par que menos afecta la información.


**Fusión de Clusters**

Se utiliza el `heap` para extraer los pares de clusters que tienen el menor `ΔI` y se realiza la fusión de dichos clusters. Esto se repite hasta alcanzar el número deseado de clusters `100`. Durante cada fusión, se recalculan las probabilidades para reflejar los cambios, y se actualiza el heap para la siguiente iteración.



**Mejoras en Complejidad:**

**Estructuras de Datos Eficientes**

Se utilizan estructuras como el `heap` para manejar los pares de clusters y sus probabilidades, lo que permite una slección rápida de un par óptimo para la fusión. Esto reduce la complejidad computacional de seleccionar pares manualmente en cada iteración.

**Técnicas de Poda**

 Al extraer elementos del heap con `heapq.heappop(heap)`, verificamos si los clusters correspondientes aún existen. Si alguno de los clusters ha sido fusionado previamente (y por lo tanto eliminado del diccionario clusters), simplemente continuamos al siguiente elemento del heap. De esta manera se realiza la poda.

In [ ]:
import heapq
import collections
import math
from typing import Dict, List, Tuple

##########
# Inicialización (Asignar cada palabra a su propio cluster)

def inicializar_clusters(vocabulario: Dict[str, int]) -> Dict[int, List[str]]:
    clusters = {i: [palabra] for palabra, i in vocabulario.items()}
    return clusters

clusters = inicializar_clusters(vocabulario)
print(f"Número inicial de clusters: {len(clusters)}")

##########
# Cálculo de probabilidades P(c)

def calcular_probabilidades_Pc(clusters: Dict[int, List[str]], corpus: List[List[str]]) -> Dict[int, float]:
    total_palabras = sum(len(doc) for doc in corpus)
    P_c = {}

    # Contar frecuencia de clases
    for c_id, palabras in clusters.items():
        frecuencia_cluster = sum(1 for doc in corpus for token in doc if token in palabras)
        P_c[c_id] = frecuencia_cluster / total_palabras

    return P_c

P_c = calcular_probabilidades_Pc(clusters, corpus_entrenamiento)
print("Probabilidades iniciales calculadas.")

##########
# Preparación de la cola de prioridad con ΔI(c_i, c_j)

def inicializar_heap(P_c: Dict[int, float], clusters: Dict[int, List[str]], corpus: List[List[str]], epsilon: float = 1e-10) -> List[Tuple[float, int, int]]:
    heap = []
    total_palabras = sum(len(doc) for doc in corpus)
    cluster_ids = list(clusters.keys())

    # Calcular P(c_i, c_j) y ΔI para pares de clusters vecinos
    for idx_i in range(len(cluster_ids)):
        c_i = cluster_ids[idx_i]
        palabras_i = set(clusters[c_i])
        for idx_j in range(idx_i + 1, len(cluster_ids)):
            c_j = cluster_ids[idx_j]
            palabras_j = set(clusters[c_j])

            # Calculamos P(c_i, c_j)
            frecuencia_conjunta = sum(
                1 for doc in corpus if any(token in palabras_i for token in doc) and any(token in palabras_j for token in doc)
            )
            if frecuencia_conjunta == 0:
                continue

            prob_conjunta = frecuencia_conjunta / total_palabras
            prob_conjunta = max(prob_conjunta, epsilon)
            P_ci = max(P_c[c_i], epsilon)
            P_cj = max(P_c[c_j], epsilon)

            delta_I = prob_conjunta * math.log(prob_conjunta / (P_ci * P_cj))

            # Añadimos al heap
            heapq.heappush(heap, (delta_I, c_i, c_j))

    return heap

heap = inicializar_heap(P_c, clusters, corpus_entrenamiento)
print("Heap inicializado con los valores de ΔI.")

##########
# Fusión de clusters con heap y poda

def fusionar_clusters_heap(clusters: Dict[int, List[str]], heap: List[Tuple[float, int, int]], P_c: Dict[int, float], corpus: List[List[str]], num_clusters_deseado: int, epsilon: float = 1e-10):
    total_palabras = sum(len(doc) for doc in corpus)

    while len(clusters) > num_clusters_deseado:
        if not heap:
            break  # No hay más fusiones posibles

        # Extraemos el par con el mínimo ΔI
        delta_I, c_i, c_j = heapq.heappop(heap)

        # Verificamos si los clusters aún existen (pueden haber sido fusionados)
        if c_i not in clusters or c_j not in clusters:
            continue

        # Fusionamos los clusters
        clusters[c_i].extend(clusters.pop(c_j))

        # Actualizamos P(c)
        P_c[c_i] += P_c.pop(c_j)

        # Actualizamos el heap con los nuevos ΔI relacionados con c_i
        for c_k in clusters:
            if c_k == c_i:
                continue

            palabras_i = set(clusters[c_i])
            palabras_k = set(clusters[c_k])

            # Calculamos P(c_i, c_k)
            frecuencia_conjunta = sum(
                1 for doc in corpus if any(token in palabras_i for token in doc) and any(token in palabras_k for token in doc)
            )
            if frecuencia_conjunta == 0:
                continue

            prob_conjunta = frecuencia_conjunta / total_palabras
            prob_conjunta = max(prob_conjunta, epsilon)
            P_ck = max(P_c[c_k], epsilon)
            P_ci = max(P_c[c_i], epsilon)

            delta_I_nuevo = prob_conjunta * math.log(prob_conjunta / (P_ci * P_ck))

            # Añadimos al heap el nuevo ΔI
            heapq.heappush(heap, (delta_I_nuevo, c_i, c_k))

        print(f"Fusión realizada. Clusters restantes: {len(clusters)}")

##########
# Número de clusters deseado (100 clusters)
num_clusters_deseado = 100

fusionar_clusters_heap(clusters, heap, P_c, corpus_entrenamiento, num_clusters_deseado)

print(f"Número final de clusters: {len(clusters)}")


Número inicial de clusters: 229
Probabilidades iniciales calculadas.
Heap inicializado con los valores de ΔI.
Fusión realizada. Clusters restantes: 228
Fusión realizada. Clusters restantes: 227
Fusión realizada. Clusters restantes: 226
Fusión realizada. Clusters restantes: 225
Fusión realizada. Clusters restantes: 224
Fusión realizada. Clusters restantes: 223
Fusión realizada. Clusters restantes: 222
Fusión realizada. Clusters restantes: 221
Fusión realizada. Clusters restantes: 220
Fusión realizada. Clusters restantes: 219
Fusión realizada. Clusters restantes: 218
Fusión realizada. Clusters restantes: 217
Fusión realizada. Clusters restantes: 216
Fusión realizada. Clusters restantes: 215
Fusión realizada. Clusters restantes: 214
Fusión realizada. Clusters restantes: 213
Fusión realizada. Clusters restantes: 212
Fusión realizada. Clusters restantes: 211
Fusión realizada. Clusters restantes: 210
Fusión realizada. Clusters restantes: 209
Fusión realizada. Clusters restantes: 208
Fusión r

**OUTPUT**

Al final, el número de clusters se reduce de 229 a 100, logrando una agrupación de términos basada en su contexto en el corpus.

La agrupación eficiente de términos basada en su contexto se logra mediante el cálculo de la pérdida de información (`ΔI`) entre cada par de clusters. Cada término se inicia en su propio cluster, y en cada iteración, el algoritmo identifica el par de clusters cuya fusión causa la menor pérdida de información, lo que implica que estos clusters comparten un contexto similar.

###  Latent Semantic Analysis (LSA)

### Latent Semantic Analysis (LSA) simple


**Construcción de la matriz término-documento con TF-IDF**

Se construye la matriz `X` donde las filas representan términos y las columnas representan documentos. Los pesos en cada posición de la matriz se calculan utilizando `TF-IDF` (Frecuencia de Término - Frecuencia Inversa de Documento), lo que captura la importancia de un término en un documento específico en relación con el corpus completo.

**Implementación de SVD usando NumPy**

Se aplica la Descomposición en valores singulares tradicional a la matriz X utilizando la función `numpy.linalg.svd`. Esta función descompone `X` en tres matrices: `U`, `S` y `Vt`. Luego, se seleccionan las primeras `k` dimensiones más significativas para reducir la dimensionalidad del espacio.

**Proyección en el espacio reducido**

Las representaciones reducidas de términos y documentos se obtienen multiplicando las matrices resultantes `U_k` y `Vt_k` por `S_k`. Esto proyecta los términos y documentos en un espacio de menor dimensión definido por las `k` componentes principales, capturando las relaciones semánticas latentes en el corpus.

In [3]:
import numpy as np
import math
from typing import List, Dict, Tuple

##########
# Construcción de la matriz término-documento con TF-IDF

def construir_matriz_termino_documento(corpus: List[List[str]], vocabulario: Dict[str, int]) -> np.ndarray:
    num_docs = len(corpus)
    num_terms = len(vocabulario)
    X = np.zeros((num_terms, num_docs))

    for j, documento in enumerate(corpus):
        term_freq = {}
        for token in documento:
            if token in vocabulario:
                idx = vocabulario[token]
                term_freq[idx] = term_freq.get(idx, 0) + 1

        for idx, freq in term_freq.items():
            tf = freq / len(documento)
            df = sum(1 for doc in corpus if vocabulario_inv[idx] in doc)
            idf = math.log((num_docs) / (df + 1)) + 1  # Suavizado
            X[idx, j] = tf * idf
    return X

# Crear el índice inverso del vocabulario
vocabulario_inv = {idx: token for token, idx in vocabulario.items()}

# Construir la matriz término-documento usando TF-IDF
X = construir_matriz_termino_documento(corpus_entrenamiento, vocabulario)
print(f"Matriz término-documento construida con tamaño {X.shape}")

##########
# Implementación de SVD usando numpy

def aplicar_svd(X: np.ndarray, k: int) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
    U, S, Vt = np.linalg.svd(X, full_matrices=False)
    # Seleccionar solo las k dimensiones significativas
    U_k = U[:, :k]
    S_k = np.diag(S[:k])
    Vt_k = Vt[:k, :]
    return U_k, S_k, Vt_k

# Selección de dimensiones significativas
k = 5
U_k, S_k, Vt_k = aplicar_svd(X, k)
print(f"SVD calculado con k={k} dimensiones significativas")

##########
# Proyección en el espacio reducido

# Obtener las representaciones reducidas para términos y documentos
representacion_terminos = U_k @ S_k
representacion_documentos = Vt_k.T @ S_k

print("\nRepresentaciones reducidas obtenidas para términos:")
print(representacion_terminos)

print("\nRepresentaciones reducidas obtenidas para documentos:")
print(representacion_documentos)


Matriz término-documento construida con tamaño (220, 1000)
SVD calculado con k=5 dimensiones significativas

Representaciones reducidas obtenidas para términos:
[[ 3.75366086e-02 -6.14277986e-01  9.75220257e-31  1.48864005e+00
  -3.01936432e+00]
 [ 9.79861628e-03 -1.47409896e-01 -4.87953903e-15  1.55969040e-01
  -1.22584378e-01]
 [ 4.04140328e-01 -7.09139110e-01 -2.16292964e-16  5.64818442e-01
  -1.77518752e+00]
 ...
 [ 7.30201345e-04 -3.03583804e-02 -2.65733526e-16  7.21819385e-02
  -2.00799277e-01]
 [ 2.20834305e-04 -1.07535447e-02  1.63927046e-16  1.84451266e-02
  -2.49167521e-02]
 [ 3.47809798e-04 -7.44887295e-03 -8.80303280e-16  2.12267513e-02
  -5.81554367e-02]]

Representaciones reducidas obtenidas para documentos:
[[ 1.85732248e-02 -9.11829054e-02 -4.16862400e-17  1.15997132e-01
  -3.07335490e-01]
 [ 4.00133757e-03 -5.27692320e-02 -6.15591836e-16  2.35633913e-02
  -7.75161100e-02]
 [ 7.58068694e-04 -4.49662729e-03 -1.29709048e-16  2.35682930e-03
  -1.61111052e-02]
 ...
 [ 8.221

### Latent Semantic Analysis (LSA) con implementaciones

**Construcción de la matriz término-documento**

Se construye la matriz `x `donde las filas representan términos y las columnas representan documentos. Los pesos asignados a cada posición en la matriz son calculados utilizando `TF-IDF`, lo que captura la importancia de un término en un documento específico en relación con el corpus.

**Implementación de SVD**

Se implementa una versión de la descomposición en valores singulares (SVD) utilizando el método de potencias. Aproxima los valores y vectores singulares más grandes de la matriz `x`. Dado que SVD completo es complejo, la aproximación nos permite obtener las primeras k componentes principales.

**Reducción de dimensionalidad**

Después de obtener las matrices `U`, `S` y `Vt` de la descomposición, se seleccionan los primeros `k` valores singulares más grandes para reducir la dimensionalidad. Las matrices `U_k` y `V_k` multiplicadas por `S_k` proporcionan las representaciones reducidas de términos y documentos.

**Proyección**

Las representaciones reducidas `representacion_terminos` y `representacion_documentos` se obtienen proyectando los términos y documentos al espacio de menor dimensión definido por las `k` componentes principales. Esto captura las relaciones semánticas en el corpus.

In [4]:
import math
import collections
from typing import List, Tuple, Dict
import random
import numpy as np

##########
# Construcción de la matriz término-documento

def construir_matriz_termino_documento(corpus: List[List[str]], vocabulario: Dict[str, int]) -> np.ndarray:
    num_docs = len(corpus)
    num_terms = len(vocabulario)
    X = np.zeros((num_terms, num_docs))

    for j, documento in enumerate(corpus):
        term_freq = {}
        for token in documento:
            idx = vocabulario[token]
            term_freq[idx] = term_freq.get(idx, 0) + 1
        # Calcular TF-IDF
        for idx, freq in term_freq.items():
            tf = freq / len(documento)
            # Calcular documento
            df = sum(1 for doc in corpus if vocabulario_inv[idx] in doc)
            idf = math.log((num_docs) / (df + 1)) + 1  # Suavizado
            X[idx, j] = tf * idf
    return X

# Construir índice inverso del vocabulario
vocabulario_inv = {idx: token for token, idx in vocabulario.items()}

# Construir la matriz término-documento
X = construir_matriz_termino_documento(corpus_entrenamiento, vocabulario)
print(f"Matriz término-documento construida con tamaño {X.shape}")

##########
# Implementación de SVD

def svd_potencias(X: np.ndarray, k: int) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
    # Implementación simplificada de SVD usando el método de potencias
    m, n = X.shape
    U = np.zeros((m, k))
    S = np.zeros((k,))
    Vt = np.zeros((k, n))

    X_copy = X.copy()
    for i in range(k):
        # Inicializar vector aleatorio
        v = np.random.rand(n)
        v = v / np.linalg.norm(v)
        # Método de potencias para obtener el valor singular máximo
        for _ in range(20):  # Número de iteraciones
            u = X_copy @ v
            u = u / np.linalg.norm(u)
            v = X_copy.T @ u
            v = v / np.linalg.norm(v)
        sigma = u.T @ X_copy @ v
        U[:, i] = u
        S[i] = sigma
        Vt[i, :] = v.T
        X_copy = X_copy - sigma * np.outer(u, v)
    return U, S, Vt

# Selección de dimensiones significativas
k = 5
U, S, Vt = svd_potencias(X, k)
print(f"SVD calculado con k={k} dimensiones significativas")

##########
# Reducción de Dimensionalidad

# Obtener las representaciones reducidas para términos y documentos
U_k_S_k = U * S  # Matriz U_k multiplicada por S_k
V_k_S_k = (Vt.T) * S  # Matriz V_k multiplicada por S_k

##########
# Proyección

# Representaciones en espacio reducido
representacion_terminos = U_k_S_k  # Representación de términos
representacion_documentos = V_k_S_k  # Representación de documentos

print("\nRepresentaciones reducidas obtenidas para términos:")
print(representacion_terminos)

print("\nRepresentaciones reducidas obtenidas para documentos:")
print(representacion_documentos)


Matriz término-documento construida con tamaño (220, 1000)
SVD calculado con k=5 dimensiones significativas

Representaciones reducidas obtenidas para términos:
[[3.75380663e-02 6.32868445e-01 5.37491415e-01 1.48632608e+00
  4.63727920e+00]
 [9.79896001e-03 1.49259495e-01 5.25386988e-02 1.49838412e-01
  1.47753779e-01]
 [4.04141977e-01 7.16052354e-01 1.93358166e-01 5.67064391e-01
  1.99467339e+00]
 ...
 [7.30273388e-04 3.12774358e-02 2.65147731e-02 7.36222428e-02
  3.16384635e-01]
 [2.20859590e-04 1.09785172e-02 6.47424089e-03 1.80519012e-02
  3.61658213e-02]
 [3.47827579e-04 7.71946581e-03 7.82719473e-03 2.16306605e-02
  9.17572577e-02]]

Representaciones reducidas obtenidas para documentos:
[[ 1.85733760e-02  9.24688803e-02  3.87060248e-02  1.14797509e-01
   3.94554752e-01]
 [ 4.00142426e-03  5.30222908e-02  7.06821257e-03  2.36107483e-02
   1.03994841e-01]
 [ 7.58076091e-04  4.52418549e-03  7.82275438e-04  2.55288855e-03
   2.31678418e-02]
 ...
 [ 8.22168475e-03  4.38547503e-02  4.3

**OUTPUT**

La matriz término-documento construida tiene un tamaño de (220,1000), donde cada fila representa un término del vocabulario y cada columna corresponde a un documento. Después de aplicar la descomposición en valores singulares (SVD) y reducir a `K=5` dimensiones significativas.

Obtenemos matrices de representaciones reducidas tanto para términos como para documentos. Las representaciones reducidas para términos y documentos nos dan un conjunto de vectores, donde cada vector describe la relación semántica de los términos y documentos con respecto a los componentes principales seleccionados.

## Word2vec

### CBOW simple

**Ventana de contexto**

El código define una ventana de contexto de tamaño c, en este caso, 2 palabras alrededor de la palabra objetivo. Esto se implementa en la función `train_cbow`, donde para cada palabra en una oración, se seleccionan los índices de las palabras de contexto en un rango que se extiende `window_size` palabras a la izquierda y a la derecha de la palabra objetivo. Las palabras de contexto se identifican excluyendo la posición de la palabra objetivo, asegurando que solo se usen palabras cercanas para calcular el contexto. Esto permite capturar la relación entre la palabra objetivo y su contexto inmediato en el corpus.

**Función de pérdida**

La función de pérdida se basa en la pérdida negativa logarítmica, implementada mediante la función sigmoide para calcular la probabilidad de que una palabra en el contexto esté relacionada con la palabra objetivo. En cada paso de entrenamiento, se calcula el producto interno entre el embedding de la palabra objetivo y el vector promedio de las palabras de contexto (`h`). Luego, se calcula el error como `1 - sigmoid` para las palabras objetivo y `0 - sigmoid_neg` para las muestras negativas. Este enfoque permite al modelo maximizar la probabilidad de co-ocurrencia entre palabras en el contexto y reducirla para palabras negativas.

**Optimización**

La optimización se realiza mediante un algoritmo básico de descenso de gradiente. En cada iteración, el vector de salida `(W_out[target]`) y el vector de entrada (`h`) se actualizan utilizando la tasa de aprendizaje (`learning_rate`) multiplicada por el error. En la etapa de "negative sampling", palabras negativas aleatorias reciben actualizaciones en función de su error, utilizando una tasa de aprendizaje ajustada. Esto permite ajustar los embeddings de manera que reflejen mejor las relaciones observadas en el contexto, de este modo se asegura que las palabras relevantes para el contexto tengan embeddings similares.

In [12]:
import numpy as np
import random

# CBOW
d = 50  # Dimension de los embeddings
window_size = 2
negative_samples = 2
epocas = 1000
learning_rate = 0.01

# Inicializamos los embeddings
np.random.seed(42)  # Para reproducibilidad
W_in = np.random.randn(len(vocabulario), d) * 0.01
W_out = np.random.randn(len(vocabulario), d) * 0.01

# Función de entrenamiento CBOW
def train_cbow(corpus, W_in, W_out, epocas=1000, learning_rate=0.01):
    for epoch in range(1, epocas + 1):
        for sentence in corpus:
            for i, palabra in enumerate(sentence):
                context_indices = list(range(max(0, i - window_size), min(len(sentence), i + window_size + 1)))
                context_indices.remove(i)  # Excluimos la palabra objetivo

                # Obtener los índices de las palabras de contexto
                context_words = [vocabulario[sentence[j]] for j in context_indices]
                target = vocabulario[palabra]

                # Vector de entrada: promedio de las palabras de contexto
                h = np.mean([W_in[context_word] for context_word in context_words], axis=0)

                # Actualizar con palabras objetivo positivas
                z = np.dot(W_out[target], h)
                sigmoid = 1 / (1 + np.exp(-z))
                error = 1 - sigmoid
                W_out[target] += learning_rate * error * h
                h += learning_rate * error * W_out[target]

                # Negative sampling
                for _ in range(negative_samples):
                    negative_word = random.randint(0, len(vocabulario)-1)
                    while negative_word == target:
                        negative_word = random.randint(0, len(vocabulario)-1)

                    z_neg = np.dot(W_out[negative_word], h)
                    sigmoid_neg = 1 / (1 + np.exp(-z_neg))
                    error_neg = 0 - sigmoid_neg
                    W_out[negative_word] += learning_rate * error_neg * h
                    h += learning_rate * error_neg * W_out[negative_word]

        # Progreso
        if epoch % 200 == 0 or epoch == 1 or epoch == epocas:
            print(f"Época {epoch} completada")

    return W_in, W_out

# Entrenamos el modelo CBOW
W_in_trained_cbow, W_out_trained_cbow = train_cbow(corpus_entrenamiento, W_in, W_out, epocas, learning_rate)

# Obtener una palabra aleatoria del vocabulario
def get_random_word(vocabulario):
    word = random.choice(list(vocabulario.keys()))
    return word

# Obtener el embedding de una palabra aleatoria del vocabulario
def get_embedding_random(W_in):
    word = get_random_word(vocabulario)
    embedding = W_in[vocabulario[word]]
    return word, embedding

# Mostrar el embedding de una palabra aleatoria
word, embedding_palabra = get_embedding_random(W_in_trained_cbow)
print(f"Embedding de '{word}': {embedding_palabra}...")


Época 1 completada
Época 200 completada
Época 400 completada
Época 600 completada
Época 800 completada
Época 1000 completada
Embedding de 'algún': [ 0.00785185 -0.01777681  0.00714746 -0.00233724  0.00707458  0.00776932
  0.01607915  0.01353454  0.01346572  0.0144367   0.00291379  0.00032185
  0.00872817  0.00732081 -0.00309066  0.00404823 -0.00035096 -0.00853933
  0.01290461  0.00055199  0.02048708 -0.01625492 -0.00556779  0.00032639
  0.01137381  0.00399469  0.00310473 -0.01127264 -0.01156941  0.00322505
 -0.01364106 -0.00531878  0.00500917 -0.00982324  0.01605356  0.01223936
 -0.0157142  -0.00932371 -0.02285777 -0.00782591  0.00766388 -0.00413895
  0.00460797 -0.01522334  0.00797975 -0.00555109  0.0001217  -0.00092875
 -0.01271549 -0.01153084]...


**OUTPUT**

La línea Embedding de `'algún'` muestra el embedding de una palabra específica del vocabulario. Este embedding es un vector de 50 dimensiones, ya que la dimensión `d` del embedding se configuró en 50 al inicio del código. Cada valor en el vector representa una dimensión de la representación semántica de la palabra `'algún'`, aprendida durante el entrenamiento. Este vector captura la relación de `'algún'` con otras palabras en el corpus en función de su contexto, permitiendo que el modelo use esta información para tareas de predicción o similitud de palabras.


### Skip-Gram con negative sampling simple

**Predicción de contexto**

En el modelo Skip-Gram, para cada palabra objetivo `(target)`, el código predice las palabras de contexto dentro de una ventana de tamaño` window_size`. Esto se logra generando un rango de índices alrededor de la palabra objetivo. La lista `context_indices` contiene los índices de las palabras dentro de la ventana de contexto, excluyendo el índice de la palabra objetivo en sí misma. Luego, el código itera sobre cada palabra en `context_indices` y la usa como contexto, calculando una predicción con el vector `W_in[target]` y el vector `W_out[context]` para ajustar sus pesos mediante descenso de gradiente.

**Negative sampling**

El código utiliza `negative sampling` que introduce palabras negativas (no en el contexto) durante el aprendizaje. En cada iteración, el código selecciona palabras negativas aleatorias de vocabulario, se asegura de que no coincidan con la palabra de contexto actual. Esta selección se realiza con la línea `neg = random.randint(0, len(vocabulario)-1)`. Luego, el código calcula el error para estas palabras negativas y ajusta los embeddings de `W_in` y `W_out` usando este error, refuerza la separación entre palabras no relacionadas.

**Función de pérdida**

La función de pérdida en este código utiliza la salida de la función sigmoide para medir la similitud entre la palabra objetivo y las palabras de contexto (pares positivos) o las palabras negativas (pares negativos). Para los pares positivos, se calcula `sigmoid = 1 / (1 + np.exp(-z))` donde `z` es el producto punto entre el embedding de la palabra objetivo y el de la palabra de contexto. El error se define como `1 - sigmoid`, y se usa para ajustar los vectores. Para las palabras negativas, el cálculo es similar, pero el error se define como `0 - sigmoid_neg`, reduce la probabilidad de relación entre la palabra objetivo y las palabras negativas.

In [13]:
import random
import numpy as np

# Parámetros
d = 50  # Dimension de los embeddings
window_size = 2
negative_samples = 2
epocas = 1000
learning_rate = 0.001

# Normalizar vectores para evitar explison de gradiente
def normalize(vec):
    norm = np.linalg.norm(vec)
    return vec / norm if norm > 0 else vec

# Inicializar los embeddings (W_in y W_out)
np.random.seed(42)
W_in = np.random.randn(len(vocabulario), d) * 0.01
W_out = np.random.randn(len(vocabulario), d) * 0.01

# Función de entrenamiento para Skip-Gram
def train_skipgram(corpus, W_in, W_out, epocas=1000, learning_rate=0.001):
    for epoch in range(1, epocas + 1):
        for sentence in corpus:
            for i, palabra in enumerate(sentence):
                target = vocabulario[palabra]
                # Definir el contexto dentro del window_size
                context_indices = list(range(max(0, i - window_size), min(len(sentence), i + window_size + 1)))
                context_indices.remove(i)  # Remover el target del contexto
                for j in context_indices:
                    context = vocabulario[sentence[j]]

                    # Positivo
                    z = np.dot(W_out[context], W_in[target])
                    sigmoid = 1 / (1 + np.exp(-z))
                    error = 1 - sigmoid
                    # Actualizar W_out y W_in para el ejemplo positivo
                    W_out[context] += learning_rate * error * W_in[target]
                    W_in[target] += learning_rate * error * W_out[context]

                    # Negativos
                    for _ in range(negative_samples):
                        neg = random.randint(0, len(vocabulario)-1)
                        while neg == context:
                            neg = random.randint(0, len(vocabulario)-1)
                        z_neg = np.dot(W_out[neg], W_in[target])
                        sigmoid_neg = 1 / (1 + np.exp(-z_neg))
                        error_neg = 0 - sigmoid_neg
                        W_out[neg] += learning_rate * error_neg * W_in[target]
                        W_in[target] += learning_rate * error_neg * W_out[neg]

                    # Normalizar vectores para evitar que se vuelvan demasiado grandes
                    W_in[target] = normalize(W_in[target])
                    W_out[context] = normalize(W_out[context])

        # Progreso
        if epoch % 200 == 0 or epoch == 1 or epoch == epocas:
            print(f"Época {epoch} completada")

    return W_in, W_out

# Entrenamos el modelo Skip-Gram con Negative Sampling
W_in_trained_sg, W_out_trained_sg = train_skipgram(corpus_entrenamiento, W_in, W_out, epocas, learning_rate)

# Obtener una palabra aleatoria del vocabulario
def get_random_word(vocabulario):
    word = random.choice(list(vocabulario.keys()))
    return word

# Obtener el embedding de una palabra aleatoria del vocabulario
def get_embedding_random(W_in):
    word = get_random_word(vocabulario)
    embedding = W_in[vocabulario[word]]
    return word, embedding

# Mostrar el embedding de una palabra
word, embedding_palabra_sg = get_embedding_random(W_in_trained_sg)
print(f"Embedding de '{word}' (Skip-Gram): {embedding_palabra_sg}...")


Época 1 completada
Época 200 completada
Época 400 completada
Época 600 completada
Época 800 completada
Época 1000 completada
Embedding de 'alemán' (Skip-Gram): [-0.22641096  0.04032551  0.08926692 -0.04716543  0.03258699  0.07874471
 -0.17789721  0.12780811 -0.12820677  0.08469878  0.13018643  0.1257329
  0.16722604  0.02180892 -0.00720884 -0.21150646 -0.03950202 -0.08072195
 -0.04290617  0.31031471  0.03337721 -0.04912377  0.16993618  0.00385315
  0.37868471 -0.19414078 -0.1400535   0.11660446  0.11522243  0.15947553
 -0.12347388  0.02654927  0.06612638  0.00372539 -0.07902223 -0.06957144
  0.0131463  -0.25117155 -0.07918373 -0.06516045 -0.002367    0.10841236
  0.12498371 -0.32530566  0.0370282  -0.03793557 -0.29913918 -0.14300072
  0.04758757  0.08380323]...


**OUTPUT**

Embedding de `'alemán'`: Después del entrenamiento, el código genera el vector de embedding de la palabra `'alemán'`. Los valores en el vector incluyen tanto números positivos como negativos, lo cual es esperado debido al uso de técnicas como negative sampling, que contribuyen a ajustar los embeddings en función de las palabras de contexto y las palabras negativas, estableciendo relaciones semánticas.

**CBOW** se centra en predecir la palabra objetivo a partir del contexto, mientras que **Skip-Gram ** intenta predecir el contexto a partir de la palabra objetivo. Ambos usan negative sampling para reducir el cálculo y mejorar el entrenamiento.

## GloVe

**Construcción de la Matriz de co-ocurrencia X**

La función `construir_matriz_coocurrencia` recorre el corpus y crea una matriz de co-ocurrencia donde cada entrada representa el número de veces que una palabra aparece en el contexto de otra dentro de una ventana determinada (tamaño `window_size`). Esta matriz captura la relación de proximidad entre palabras en el corpus.

**Inicialización de vectores y sesgos**

 Los vectores de palabras (`W`) y de contexto (`W_contexto`) se inicializan aleatoriamente utilizando una distribución normal. Además, se inicializan los sesgos `b` y `b_contexto` como vectores de ceros para cada palabra en el vocabulario.

**Optimización de la función de costo**

La función `entrenar_glove` implementa el entrenamiento del modelo GloVe mediante una función de costo que minimiza la diferencia entre el producto escalar de los vectores de palabra y contexto, más sus sesgos, y el logaritmo de las frecuencias de co-ocurrencia. Se utiliza un gradiente descendente estocástico para actualizar los vectores y sesgos en cada iteración, con el objetivo de minimizar el error.

**Iteración hasta convergencia**

El entrenamiento se realiza durante un número predefinido de épocas (`epocas`), actualizando los parámetros hasta que se alcanza un mínimo del error total. Durante el proceso, el error total se imprime en ciertas épocas para registrar la convergencia del modelo y asegurarse de que la función de costo está disminuyendo.

In [17]:
import numpy as np

# GloVe
d = 50  # Dimensión de los embeddings
epocas = 350
learning_rate = 0.05
x_max = 100  # Parámetro de suavizado
alpha = 0.75  # Exponente en la función de ponderación

# Construcción de la matriz de co-ocurrencia
def construir_matriz_coocurrencia(corpus, vocabulario, window_size=2):
    matriz_coocurrencia = np.zeros((len(vocabulario), len(vocabulario)))
    for documento in corpus:
        for i, palabra in enumerate(documento):
            palabra_indice = vocabulario[palabra]
            # Contexto a la izquierda y derecha
            for j in range(max(i - window_size, 0), min(i + window_size + 1, len(documento))):
                if i != j:
                    contexto_indice = vocabulario[documento[j]]
                    matriz_coocurrencia[palabra_indice, contexto_indice] += 1
    return matriz_coocurrencia

# Inicialización de vectores y sesgos
np.random.seed(42)
W = np.random.randn(len(vocabulario), d) * 0.01  # Vectores de palabras (objetivo)
W_contexto = np.random.randn(len(vocabulario), d) * 0.01  # Vectores de contexto
b = np.zeros(len(vocabulario))  # Sesgos para palabras objetivo
b_contexto = np.zeros(len(vocabulario))  # Sesgos para palabras de contexto

# Función de ponderación
def funcion_ponderacion(x, x_max, alpha):
    return (x / x_max)**alpha if x < x_max else 1

# Entrenamiento de GloVe
def entrenar_glove(matriz_coocurrencia, W, W_contexto, b, b_contexto, epocas, learning_rate, x_max, alpha):
    for epoch in range(epocas):
        total_error = 0
        # Iterar sobre cada par de palabras
        for i in range(len(vocabulario)):
            for j in range(len(vocabulario)):
                if matriz_coocurrencia[i, j] > 0:
                    X_ij = matriz_coocurrencia[i, j]
                    # Calcular el error y la función de pérdida
                    w_dot = np.dot(W[i], W_contexto[j])
                    log_Xij = np.log(X_ij)
                    peso = funcion_ponderacion(X_ij, x_max, alpha)
                    error = (w_dot + b[i] + b_contexto[j] - log_Xij)
                    total_error += peso * (error ** 2)

                    # Gradientes
                    grad_w = peso * error * W_contexto[j]
                    grad_w_contexto = peso * error * W[i]
                    grad_b = peso * error
                    grad_b_contexto = peso * error

                    # Actualizar los parámetros
                    W[i] -= learning_rate * grad_w
                    W_contexto[j] -= learning_rate * grad_w_contexto
                    b[i] -= learning_rate * grad_b
                    b_contexto[j] -= learning_rate * grad_b_contexto

        # Progreso
        if (epoch + 1) % 100 == 0 or epoch == 0:
            print(f"Época {epoch + 1}/{epocas}, Error total: {total_error:.4f}")

    return W, W_contexto

# Construir matriz de co-ocurrencia
matriz_coocurrencia = construir_matriz_coocurrencia(corpus_entrenamiento, vocabulario)

# Entrenar el modelo GloVe
W_trained, W_contexto_trained = entrenar_glove(matriz_coocurrencia, W, W_contexto, b, b_contexto, epocas, learning_rate, x_max, alpha)

# Obtener el embedding de una palabra aleatoria del vocabulario
def get_embedding_random(W):
    word = random.choice(list(vocabulario.keys()))
    embedding = W[vocabulario[word]]
    return word, embedding

# Mostrar el embedding de una palabra aleatoria
word, embedding_palabra = get_embedding_random(W_trained)
print(f"Embedding de '{word}' (GloVe): {embedding_palabra}...")


Época 1/350, Error total: 291.0738
Época 100/350, Error total: 44.3040
Época 200/350, Error total: 19.9556
Época 300/350, Error total: 10.5266
Embedding de 'social' (GloVe): [ 0.09853697  0.06702795  0.14756727 -0.07771305  0.28116965 -0.03490453
  0.17554624 -0.16920621  0.12402475 -0.22440128  0.15073892 -0.28913077
  0.13841251  0.2651067  -0.12441293  0.16481013 -0.07080443 -0.02971872
  0.1696008  -0.10228858 -0.01757578 -0.11689146 -0.10164319 -0.08348371
 -0.42876538  0.1207345   0.04650726 -0.04542867 -0.11074363 -0.23176622
 -0.03405249 -0.34122812 -0.13890086 -0.16995003 -0.02052767  0.19202377
  0.11511954 -0.05574901  0.17863603 -0.25131119 -0.01812615 -0.04267015
 -0.16474377  0.43447716 -0.10171589  0.19114515  0.06633759  0.27699309
 -0.17502346 -0.02310334]...


**OUTPUT**

Inicialmente, el error total es alto (291.07 en la primera época), pero disminuye a 10.52 en la época 300, mostrando una convergencia hacia valores más bajos.

El último mensaje muestra el embedding generado para la palabra `'social'`. Este vector tiene 50 dimensiones (`d = 50`), y cada valor representa una característica que captura relaciones semánticas con otras palabras. Los valores incluyen tanto positivos como negativos, reflejando una combinación de interacciones en el contexto de la palabra `'social'`.

# Parte 4

### NO IMPLEMENTADO